### PR TITLE
fix: nginx setup for deployment with ssl

### LINKED ISSUE #1157: cannot deploy to new VM
**Describe the bug**

Deploying to the new VM leads to errors.

**Steps to reproduce**

After adding the Docker context, and updating the environment, the deployment command succeeds but the `nginx` container restarts in a loop.

**Expected behavior**

- [ ] a successful deployment

**Screenshots**

**Additional context**

### PR DESCRIPTION
**Pull request by mihai-sysbio**
**Related issue(s) and PR(s)**  
This PR closes #1157 and relates to https://github.com/MetabolicAtlas/private-issues/issues/38.

**Type of change**  

- [x] Bug fix (non-breaking change which fixes an issue)

**List of changes made**  

The production Dockerfile was updated, and an Nginx configuration was added to disable IPv6 (see comments below for full story).

**Testing**  

- Instructions on how to test: deploy the stack as usual to either `dev` or `prod3`

**Further comments**  

This was also researched by @MalinAhlberg and @inghylt.

In the end, there were multiple problems, mainly stemming from using Let's Encrypt certificates though a custom Docker image. This is marked for removal in https://github.com/MetabolicAtlas/private-issues/issues/38 . However, the FTP server still requires the certificates, so this functionality can only be removed with the removal of the FTP server.

The certificates can be obtained from the current container, but port has to be manually changes from `1337` to `80`. Opening the port up in Docker compose did not help, and neither did the VM's iptables configuration.

After obtaining the certificates manually, these are stored in a permanent volume, so normally the problem should have gone away. However, the container kept restarting because of a new problem having to do with how the `nginx` service was restarted https://github.com/staticfloat/docker-nginx-certbot-old/blob/5d2eef666c639fcaadc237492a46c4371d392cb1/src/scripts/entrypoint.sh#L41

This also pointed out that the image we were using was old, so the next step was to switch to a new one https://hub.docker.com/r/jonasal/nginx-certbot . As described in the [README](https://github.com/JonasAlfredsson/docker-nginx-certbot#acknowledgments-and-thanks), this image is fork of the fork from the same original author. So, it made sense to switch back to it.

This new image however also enables IPv6, which had to be disabled by fetching the original file, commenting out the IPv6 line which is not supported by our VM, and overwriting the original one in the container. Moreover, this new image drops support of templating the way it was previously done, so a new approach was taken using the native Nginx templating https://github.com/JonasAlfredsson/docker-nginx-certbot/blob/master/docs/good_to_know.md#help-migrating-from-staticfloats-image .

**Definition of Done checklist**  
- [x] My code meets the Acceptance Criteria
- [x] My code follows the [NBIS style guidelines](https://github.com/NBISweden/development-guidelines)
- [x] I have performed a self-review of my own code and commented any hard-to-understand areas
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by JonasAlfredsson** [comment]
Hi mihai-sysbio,

I just happened to stumble upon this pull request and saw that you are about to migrate to my nginx-certbot image, if you have any questions or issues please don't hesitate to reach out and I will see if I can assist :)

However, the reason I write here is because of this comment:

> the container kept restarting because of a new problem having to do with how the nginx service was restarted https://github.com/staticfloat/docker-nginx-certbot-old/blob/5d2eef666c639fcaadc237492a46c4371d392cb1/src/scripts/entrypoint.sh#L41

I have seen this issue being mentioned on some other places and would like to ask how the problem affects you (for example are you deploying to elastic beanstalk or similar), and if it is solved with how I reload Nginx?

**Comment by mihai-sysbio** [comment]
> Hi mihai-sysbio,
> 
> I just happened to stumble upon this pull request and saw that you are about to migrate to my nginx-certbot image, if you have any questions or issues please don't hesitate to reach out and I will see if I can assist :)

Hi @JonasAlfredsson! It's very nice of you to reach out - so far things seem to be working fine. One suggestion I could come up with is the thing mentioned above with disabling IPv6 (or IPv4 for that matter), it might be nice to have that documented in an faq section.

> However, the reason I write here is because of this comment:
> 
> > the container kept restarting because of a new problem having to do with how the nginx service was restarted https://github.com/staticfloat/docker-nginx-certbot-old/blob/5d2eef666c639fcaadc237492a46c4371d392cb1/src/scripts/entrypoint.sh#L41
> 
> I have seen this issue being mentioned on some other places and would like to ask how the problem affects you (for example are you deploying to elastic beanstalk or similar), and if it is solved with how I reload Nginx?

We are deploying to our own VMs. Previously, the error was that there was no such Nginx PID, so it couldn't kill Nginx in order to (re)start it. So far, we haven't had any issues, but it might also be because the certificate generation has not yet failed. I'll remember to reach out if any issue occurs.

**Comment by nanjiangshu** [comment]
I thought we are going to replace the third party nginx+certbot image with the official `nginx` image and then use Cloudfare to setup the SSL. However, it is still a third party image is used. I'm not sure if the image `jonasal/nginx-certbot` is better. When I tested on chemical space with this image, it seems the variable `ENVSUBST_VARS` is not working so that the `SERVER_NAME` will not be substituted if that is set up on the Dockerfile.

**Comment by JonasAlfredsson** [comment]
> This is my last comment here unless you mention me directly, as I don't want to interfere with your discussion here.

@mihai-sysbio, I thought Nginx just ignored enabling IPv6 if it is not present (and IPv6 is not present in docker unless activated). Please open an issue in the repo if this is something that you would want supported. :)

@nanjiangshu, my image is based on the official Nginx image, and it is rebased as soon as the official Nginx docker repo publishes new images. I try to have my code being as small component as possible in the image.
Also, depending on how you intend to use cloudflare SSL you will probably want the connection between [them and your server](https://community.cloudflare.com/t/ssl-tls-app-settings/53186) to be secure as well, so you will probably need some kind of certs locally. For this you can go the semi-advanced route of having [your own CA](https://github.com/JonasAlfredsson/docker-nginx-certbot/blob/master/docs/advanced_usage.md#local-ca).

**Review by e0 — APPROVED** [review]
Good job! Nice that we got everything sorted out.

Deploying to the dev server seems to work. I had to clean up some spaces and restart the `neo4j` and `pg` containers but that's an unrelated problem. Just wanted to mention again for anyone else that's trying to deploy.

**Comment by e0** [comment]
Thank you for your work on the image and taking time to help us @JonasAlfredsson! Your image seems to work great for us and these are really good tips.

This project is a bit unique in the sense that there is no user specific session and all data is already publicly available. Having a user facing SSL certificate is only to be compliant with browsers. We figured it is probably fine to not have the SSL between Cloudflare to our server. However there might be more reasons to have the SSL and we (the team) could probably do some more research here.

Your image also seems to be in much better shape in quality and maintenance compared to what we were using before. Since it is such as lean layer on top of the official `nginx` image and we still do have a need for the certbot (for our FTP server), I think your image is really ideal for us. Thanks again!

**Comment by JonasAlfredsson** [comment]
@e0, nice to hear that it works, if something breaks do be afraid to reach out and I will see what I can do to assist :)

In regards to not encrypting your traffic: I think it is absolutely bonkers to send anything over the internet without encryption today. Even though "everything is already publicly available", HTTPS helps users verify that the data they receive actually comes from you and that it has not been tampered with, and that is something I would want my users to know. 
I always question the seriousness of sites which don't support HTTPS when I browse around, so without wanting to sound rude I would say you absolutely say you need it between you and Cloudflare :)

**Comment by e0** [comment]
@JonasAlfredsson it's not rude at all and I really appreciate your sincere input! It sounds like I got some basic cybersecurity stuff to read about SSL.

**Comment by nanjiangshu** [comment]
@JonasAlfredsson Thanks very much for providing with such as nice and well maintained image and great to know that your image up-to-date with the office nginx image. I agree with that the even the communication between our server and the Cloudfare should be protected. I was just a bit unclear (at the beginning) for which image to use within out team, a third party image or the nginx office image. Since you are so kind and willingly to discuss with us regarding your image, could you please confirm that whether the variables defined in `ENVSUBST_VARS` will be replaced by their actually values once the containers are created?

**Comment by JonasAlfredsson** [comment]
@nanjiangshu, I think you right in wanting to use as few "third party" images as possible, for what you know I am just some random dude on the internet :). However, if you don't specify paths to certificate files in your config files my image is basically identical to the official Nginx.

In regards to the `ENVSUBST_VARS`, I am unsure where this variable comes from. 
If you are looking for variable substitution the offical Nginx image does it like [this](https://github.com/docker-library/docs/tree/master/nginx#using-environment-variables-in-nginx-configuration-new-in-119):

1. At restart [this entrypoint](https://github.com/nginxinc/docker-nginx/blob/master/entrypoint/20-envsubst-on-templates.sh) script is run
2. All files matching `/etc/nginx/templates/<name>.template` is run through `envsubst`
3. If you set `-e NGINX_PORT=80` the line `listen ${NGINX_PORT};` will be transformed to `listen 80;`
4. The resulting files end up in `/etc/nginx/conf.d/<name>`

File: nginx/prod.Dockerfile [hunk 1]
```diff
@@ -7,6 +7,7 @@ WORKDIR /project
 COPY frontend .
 RUN yarn && yarn build
 
-FROM staticfloat/nginx-certbot
-COPY nginx/prod.nginx.conf /etc/nginx/user.conf.d/nginx.conf
+FROM jonasal/nginx-certbot:3.2.0-nginx1.23.1-alpine
+COPY nginx/redirector.conf /etc/nginx/conf.d/redirector.conf
+COPY nginx/prod.nginx.conf /etc/nginx/templates/default.conf.template
```
Comment (mihai-sysbio) [inline nginx/prod.Dockerfile:12]: > 2\. All files matching `/etc/nginx/templates/<name>.template` is run through `envsubst`

This is used in the current PR in the line above.

**Comment by nanjiangshu** [comment]
@JonasAlfredsson Thanks very much for clarification!

File: nginx/prod.Dockerfile (hunk 1 above)
Comment (nanjiangshu) [reply nginx/prod.Dockerfile:12]: Great! Then maybe this line https://github.com/MetabolicAtlas/MetabolicAtlas/blob/main/docker-compose-remote.yml#L18 is not needed anymore?

Comment (mihai-sysbio) [reply nginx/prod.Dockerfile:12]: Excellent point @nanjiangshu, that line is now removed via [2e5455f](https://github.com/MetabolicAtlas/MetabolicAtlas/pull/1158/commits/2e5455f75cc5d5dfaa02db26bc8a6dab5eaf66b1).

**Review by nanjiangshu — APPROVED** [review]
It worked well when tried on the `dev` server. 

However, the images on the about page are still missing https://dev-metatlas.csbi.chalmers.se/about/platform#advisory-board. This problem is not related to the change of nginx image and seems happen only when deploying from Linux. I will create a new issue for that.

**Pull request merged**